pkg/yurthub/filter/nodeportisolation/filter.go

```diff
@@ -0,0 +1,238 @@
+		nif.nodeSynced = factory.Core().V1().Nodes().Informer().HasSynced
+		nif.nodeGetter = factory.Core().V1().Nodes().Lister().Get
+	}
+
+	return nil
+}
+
+func (nif *nodePortIsolationFilter) SetStorageWrapper(s cachemanager.StorageWrapper) error {
+	if s.Name() != disk.StorageName {
+		return fmt.Errorf("nodePortIsolationFilter can only support disk storage currently, cannot use %s", s.Name())
+	}
+
+	if len(nif.nodeName) == 0 {
+		return fmt.Errorf("node name for nodePortIsolationFilter is not ready")
+	}
+
+	// hub agent will list/watch node from kube-apiserver when hub agent work as cloud mode
+	if nif.workingMode == util.WorkingModeCloud {
+		return nil
+	}
+	klog.Infof("prepare local disk storage to sync node(%s) for edge working mode", nif.nodeName)
+
+	nodeKey, err := s.KeyFunc(storage.KeyBuildInfo{
+		Component: "kubelet",
+		Name:      nif.nodeName,
+		Resources: "nodes",
+		Group:     "",
+		Version:   "v1",
+	})
+	if err != nil {
+		return fmt.Errorf("failed to get node key for %s, %v", nif.nodeName, err)
+	}
+	nif.nodeSynced = func() bool {
+		obj, err := s.Get(nodeKey)
+		if err != nil || obj == nil {
+			return false
+		}
+
+		if _, ok := obj.(*v1.Node); !ok {
+			return false
+		}
+
+		return true
+	}
+
+	nif.nodeGetter = func(name string) (*v1.Node, error) {
+		key, err := s.KeyFunc(storage.KeyBuildInfo{
+			Component: "kubelet",
+			Name:      name,
+			Resources: "nodes",
+			Group:     "",
+			Version:   "v1",
+		})
+		if err != nil {
+			return nil, fmt.Errorf("nodeGetter failed to get node key for %s, %v", name, err)
+		}
+		obj, err := s.Get(key)
+		if err != nil {
+			return nil, err
+		} else if obj == nil {
+			return nil, fmt.Errorf("node(%s) is not ready", name)
+		}
+
+		if node, ok := obj.(*v1.Node); ok {
+			return node, nil
+		}
+
+		return nil, fmt.Errorf("node(%s) is not found", name)
+	}
+
+	return nil
+}
+
+func (nif *nodePortIsolationFilter) Filter(obj runtime.Object, stopCh <-chan struct{}) runtime.Object {
+	if ok := cache.WaitForCacheSync(stopCh, nif.nodeSynced); !ok {
+		return obj
+	}
+
+	switch v := obj.(type) {
+	case *v1.ServiceList:
+		var svcNew []v1.Service
+		for i := range v.Items {
+			svc := nif.isolateNodePortService(&v.Items[i])
+			if svc != nil {
+				svcNew = append(svcNew, *svc)
+			}
+		}
+		v.Items = svcNew
+		return v
+	case *v1.Service:
+		return nif.isolateNodePortService(v)
+	default:
+		return v
+	}
+}
+
+func (nif *nodePortIsolationFilter) isolateNodePortService(svc *v1.Service) *v1.Service {
+	nodePoolName := nif.nodePoolName
+	if len(nodePoolName) == 0 {
+		node, err := nif.nodeGetter(nif.nodeName)
+		if err != nil {
+			klog.Warningf("skip isolateNodePortService filter, failed to get node(%s), %v", nif.nodeName, err)
+			return svc
+		}
+		nodePoolName = node.Labels[nodepoolv1alpha1.LabelCurrentNodePool]
+	}
+
+	// node is not located in NodePool, skip this filter
+	if len(nodePoolName) == 0 {
+		return svc
```

Comment:
Is it suitable that skipping filter for node that does not in any nodepool?

1. If users only set enabled nodepools for svcs, it looks strange that such nodePort svcs will also work on these orphan nodes.

2. But if users only set disabled nodepools for svcs, it's intuitive that these svcs should work on all nodes in other nodepools and all orphan nodes. 

3. If users set both enable nodepools and disable nodepools, it maybe confused because such two sets may overlap. For example, "np1,np2,-np1,-np3", we have to make more clear definition for such case.

Reply:
I have updated nodepool definition in #1183 as following:

The details of nodeport.openyurt.io/isolation annotation of service:

key: nodeport.openyurt.io/isolation

value: you can combine [np, -np, *] these three string to configure value, the priority is: [ np > -np > * ]
- np: enable specified NodePort service listening in this nodePool
- -np: disable specified NodePort service listening in this nodePool
- *: enable specified NodePort service listening in all nodePools except specified by -np.

for example:

*: enable NodePort service listening in all nodePools
nodepool1, nodepool2: enable NodePort service listening in nodePool1 and nodePool2
-nodepool1, *: enable NodePort service listening in all nodePools except nodePool1
nodepool1,-nodepool2: only enable NodePort service listening in nodePool1
nodepool1, *: enable NodePort service listening in all nodePools
nodepool1, -nodepool1, *: enable NodePort service listening in of nodePools except nodePool1

Reply:
> nodepool1, -nodepool1, *: enable NodePort service listening in of nodePools except nodePool1

If [ np > -np > * ], I think it should enable NodePort service on all nodepools.

Is there a better way to indicate orphan nodes? Using `!` is hard to read I think.

Reply:
@Congrool As discussed in community meeting, the design has been changed as following:

I will name the new filter as `nodeportisolation`, and add a new annotation `nodeport.openyurt.io/listen` for NodePort service in order to trigger `nodeportisolation` filter in Yurthub.

The details of `nodeport.openyurt.io/listen` annotation of service:
- key:    `nodeport.openyurt.io/listen`
- value:  A list of nodepool names that separated by ','
   *  `foo`:  enable specified NodePort service listening on nodes of NodePool named foo
   * `-foo`: disable specified NodePort service listening on nodes of NodePool named fool
   * `*`: enable specified NodePort service listening on nodes of all NodePools
   * if NodePool name is duplicated in the configuration, we will take the first configuration
   * if NodePool name is not configured, we will disable NodePort listening on nodes of these unconfigured NodePools.
   * orphan nodes(don't locate in NodePool) will be kept the same as native Kubernetes, so NodePort will be listened on orphan nodes.

     for example:
     - `foo, bar`:  enable NodePort service listening on nodes in foo and bar NodePool.
     - `foo, *`: enable NodePort service listening on nodes of all NodePools
     - `-foo, -bar`: disable NodePort service listening on nodes of all NodePools
     - `-foo, *`: disable NodePort service listening only on nodes in foo NodePool
     - `foo,-foo`: enable NodePort service listening on nodes in foo NodePool
     - `-foo`: disable NodePort service listening on nodes of all NodePools(include foo NodePool).